### PR TITLE
EZP-31450: Introduced strict types for URLAliasService

### PR DESCRIPTION
**Pull request by adamwojs**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31450](https://jira.ez.no/browse/EZP-31450)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** |  `master`
| **BC breaks**      | yes
| **Tests pass**     | yes
| **Doc needed**     | yes

This PR introduces strict types for the \eZ\Publish\API\Repository\URLAliasService

**TODO**:
- [X] Implement feature / fix a bug.
- [X] Implement tests.
- [X] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/API/Repository/URLAliasService.php [hunk 1]
```diff
@@ -60,29 +65,43 @@ public function createUrlAlias(Location $location, $path, $languageCode, $forwar
...
      *
      * @return \eZ\Publish\API\Repository\Values\Content\URLAlias[]
      */
-    public function listLocationAliases(Location $location, $custom = true, $languageCode = null);
+    public function listLocationAliases(
+        Location $location,
+        bool $custom = true,
+        ?string $languageCode = null,
+        bool $showAllTranslations = null,
+        array $prioritizedLanguageList = null
```
Comment (webhdx) [inline eZ/Publish/API/Repository/URLAliasService.php]: If this is intended to be an array then why it is nullable? Can't you use an empty array as a default value?

File: eZ/Publish/Core/FieldType/Validator/ImageValidator.php [hunk 2]
```diff
@@ -13,15 +13,15 @@
 class ImageValidator extends Validator
 {
     /**
-     * @inheritDoc
+     * {@inheritdoc}
```
Comment (webhdx) [inline eZ/Publish/Core/FieldType/Validator/ImageValidator.php]: Is it really needed? I think most tools should already use docblock from parent method.

File: eZ/Publish/Core/FieldType/Validator/ImageValidator.php [hunk 3]
```diff
@@ -13,15 +13,15 @@
...
      */
     public function validateConstraints($constraints)
     {
         return [];
     }
 
     /**
-     * @inheritDoc
+     * {@inheritdoc}
```
Comment (webhdx) [inline eZ/Publish/Core/FieldType/Validator/ImageValidator.php]: Same as above

File: eZ/Publish/API/Repository/URLAliasService.php [hunk 4]
```diff
@@ -118,11 +137,18 @@ public function lookup($url, $languageCode = null);
...
+     * @param string[]|null $prioritizedLanguages
      *
      * @return \eZ\Publish\API\Repository\Values\Content\URLAlias
      */
-    public function reverseLookup(Location $location, $languageCode = null);
+    public function reverseLookup(
+        Location $location,
+        ?string $languageCode = null,
+        ?bool $showAllTranslations = null,
+        ?array $prioritizedLanguages = null
```
Comment (webhdx) [inline eZ/Publish/API/Repository/URLAliasService.php]: `listLocationAliases` uses `array $prioritizedLanguages = null` (not nullable with `null` as default value). This should be streamlined but for me they both should be not nullable and use an empty array as a default value.

File: eZ/Publish/API/Repository/URLAliasService.php [hunk 5]
```diff
@@ -118,11 +137,18 @@ public function lookup($url, $languageCode = null);
...
+     * @param bool|null $showAllTranslations
+     * @param string[]|null $prioritizedLanguages
      *
      * @return \eZ\Publish\API\Repository\Values\Content\URLAlias
      */
-    public function reverseLookup(Location $location, $languageCode = null);
+    public function reverseLookup(
+        Location $location,
+        ?string $languageCode = null,
+        ?bool $showAllTranslations = null,
```
Comment (webhdx) [inline eZ/Publish/API/Repository/URLAliasService.php]: Same as below but - I don't think this should be nullable.

File: eZ/Publish/Core/FieldType/Validator/ImageValidator.php (hunk 2 above)
Comment (adamwojs) [reply eZ/Publish/Core/FieldType/Validator/ImageValidator.php]: This change was not intended.

File: eZ/Publish/Core/FieldType/Validator/ImageValidator.php (hunk 3 above)
Comment (adamwojs) [reply eZ/Publish/Core/FieldType/Validator/ImageValidator.php]: This change was not intended.

File: eZ/Publish/API/Repository/URLAliasService.php [hunk 6]
```diff
@@ -118,11 +137,18 @@ public function lookup($url, $languageCode = null);
      * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException if no url alias exist for the given language
      *
      * @param \eZ\Publish\API\Repository\Values\Content\Location $location
-     * @param string $languageCode
+     * @param string|null $languageCode
+     * @param bool|null $showAllTranslations
+     * @param string[]|null $prioritizedLanguages
```
Comment (alongosz) [inline eZ/Publish/API/Repository/URLAliasService.php]: @adamwojs if you're touching PHPDocs in the scope, then you should drop all attributes which don't provide any extra information with respect to the method prototype (use [inspections profile](https://github.com/ezsystems/ezplatform-code-style/pull/1/files) to see what really is needed (e.g. `array` is considered non-strict and needs to be documented)

File: eZ/Publish/API/Repository/URLAliasService.php (hunk 1 above)
Comment (adamwojs) [reply eZ/Publish/API/Repository/URLAliasService.php]: > Can't you use an empty array as a default value?

Unfortunately no. `null` is an indicator for SA layer that `$showAllTranslations` and `$prioritizedLanguageList` value should be resolved base on current site access settings.  An empty array means "all languages" (ref. `\eZ\Publish\API\Repository\Values\Content\Language::ALL`). Any other values are used as-is. 

It's not the best API design but PR is not intended to change the behavior of any of refactored methods.

**Comment by adamwojs** [comment]
PR updated according to @webhdx suggestions.

**Comment by katarzynazawada** [comment]
QA Approved
I found 2 issues during testing but they occur albo on actual master 
https://jira.ez.no/browse/EZP-31490
https://jira.ez.no/browse/EZP-31492

**Pull request merged**